src/fifelse.c

```diff
@@ -201,105 +201,106 @@ SEXP fifelseR(SEXP l, SEXP a, SEXP b, SEXP na) {
   return ans;
 }
 
-SEXP fcaseR(SEXP na, SEXP rho, SEXP args) {
-  const int narg=length(args);
+SEXP fcaseR(SEXP rho, SEXP args) {
+  const int narg=length(args); // `default` will take the last two positions
   if (narg % 2) {
     error(_("Received %d inputs; please supply an even number of arguments in ..., "
             "consisting of logical condition, resulting value pairs (in that order). "
-            "Note that the default argument must be named explicitly, e.g., default=0"), narg);
+            "Note that the default argument must be named explicitly, e.g., default=0"), narg - 2);
   }
-  if (narg==0) return R_NilValue;
-
-  SEXP cons0 = PROTECT(eval(SEXPPTR_RO(args)[0], rho));
-  SEXP value0 = PROTECT(eval(SEXPPTR_RO(args)[1], rho)); // value0 will be compared to from loop so leave it protected throughout
-  SEXPTYPE type0 = TYPEOF(value0);
-  int64_t len0=xlength(cons0), len2=len0;
-  if (isS4(value0) && !INHERITS(value0, char_nanotime)) {
-    error(_("S4 class objects (except nanotime) are not supported. Please see https://github.com/Rdatatable/data.table/issues/4131."));
-    // otherwise 'invalid type/length (S4/1) in vector allocation' from test 2132.3
-  }
-  SEXP ans = PROTECT(allocVector(type0, len0));
-  SEXP tracker = PROTECT(allocVector(INTSXP, len0));
-  int *restrict p = INTEGER(tracker);
-  copyMostAttrib(value0, ans);
-
-  bool nonna=!isNull(na);
-  if (nonna) {
-    if (xlength(na) != 1) {
-      error(_("Length of 'default' must be 1."));
-    }
-    SEXPTYPE tn = TYPEOF(na);
-    if (tn==LGLSXP && LOGICAL(na)[0]==NA_LOGICAL) {
-      nonna = false;
-    } else {
-      if (tn != type0) {
-        error(_("Resulting value is of type %s but 'default' is of type %s. "
-                "Please make sure that both arguments have the same type."), type2char(type0), type2char(tn));
-      }
-      if (!R_compute_identical(PROTECT(getAttrib(value0,R_ClassSymbol)),  PROTECT(getAttrib(na,R_ClassSymbol)), 0)) {
-        error(_("Resulting value has different class than 'default'. "
-                "Please make sure that both arguments have the same class."));
-      }
-      UNPROTECT(2);
-      if (isFactor(value0)) {
-        if (!R_compute_identical(PROTECT(getAttrib(value0,R_LevelsSymbol)),  PROTECT(getAttrib(na,R_LevelsSymbol)), 0)) {
-          error(_("Resulting value and 'default' are both type factor but their levels are different."));
-        }
-        UNPROTECT(2);
-      }
-    }
-  }
-
+  int nprotect=0, l=0;
+  int64_t len0=0, len1=0, len2=0, idx=0;
```

Comment:
idx is never used below but always masekd with `const int64_t idx`
```suggestion
  int64_t len0=0, len1=0, len2=0;
```